pjlib-util/src/pjlib-util/upnp.c

```diff
@@ -0,0 +1,907 @@
+         * from port mapping deletion action.
+         */
+        err_code = UpnpActionComplete_get_ErrCode(a_event);
+        if (err_code != UPNP_E_SUCCESS) {
+            PJ_LOG(4, (THIS_FILE, "Port mapping deletion action complete "
+            			  "error: %d (%s)", err_code,
+            			  UpnpGetErrorMessage(err_code)));
+            break;
+        }
+        
+        response = UpnpActionComplete_get_ActionResult(a_event);
+        if (!response) {
+            PJ_LOG(4, (THIS_FILE, "Failed to get response to delete port "
+            			  "mapping"));
+        } else {
+            if (!check_error_response(response)) {
+        	PJ_LOG(4, (THIS_FILE, "Successfully deleted port mapping"));
+            }
+            ixmlDocument_free(response);
+        }
+
+        break;
+    }
+    default:
+        TRACE_("Unhandled UPnP client callback %d", event_type);
+        break;
+    }
+    
+    return UPNP_E_SUCCESS;
+}
+
+/* Initiate search for Internet Gateway Devices. */
+static void search_igd(int search_time)
+{
+    int err;
+
+    upnp_mgr.search_cnt = 4;
+
+    err = UpnpSearchAsync(upnp_mgr.client_hnd, search_time,
+    			  UPNP_ROOT_DEVICE, &upnp_mgr);
+    if (err != UPNP_E_SUCCESS) {
+        PJ_LOG(4, (THIS_FILE, "Searching for UPNP_ROOT_DEVICE failed: %s",
+                  	      UpnpGetErrorMessage(err)));
+    }
+
+    err = UpnpSearchAsync(upnp_mgr.client_hnd, search_time,
+    			  UPNP_IGD_DEVICE, &upnp_mgr);
+    if (err != UPNP_E_SUCCESS) {
+        PJ_LOG(4, (THIS_FILE, "Searching for UPNP_IGD_DEVICE failed: %s",
+                  	      UpnpGetErrorMessage(err)));
+    }
+
+    err = UpnpSearchAsync(upnp_mgr.client_hnd, search_time,
+    			  UPNP_WANIP_SERVICE, &upnp_mgr);
+    if (err != UPNP_E_SUCCESS) {
+        PJ_LOG(4, (THIS_FILE, "Searching for UPNP_WANIP_SERVICE failed: %s",
+                  	      UpnpGetErrorMessage(err)));
+    }
+
+    err = UpnpSearchAsync(upnp_mgr.client_hnd, search_time,
+    			  UPNP_WANPPP_SERVICE, &upnp_mgr);
+    if (err != UPNP_E_SUCCESS) {
+        PJ_LOG(4, (THIS_FILE, "Searching for UPNP_WANPPP_SERVICE failed: %s",
+                  	      UpnpGetErrorMessage(err)));
+    }
+}
+
+/* Initialize UPnP. */
+PJ_DEF(pj_status_t) pj_upnp_init(const pj_upnp_init_param *param)
+{
+    int upnp_err;
+    const char *ip_address;
+    unsigned short port;
+    const char *ip_address6 = NULL;
+    unsigned short port6 = 0;
+    pj_status_t status;
+
+    if (upnp_mgr.initialized)
+    	return PJ_SUCCESS;
+
+#if ENABLE_LOG
+    UpnpSetLogLevel(UPNP_ALL);
+    UpnpSetLogFileNames("upnp.log", NULL);
+    upnp_err = UpnpInitLog();
+    if (upnp_err != UPNP_E_SUCCESS) {
+	PJ_LOG(4, (THIS_FILE, "Failed to initialize UPnP log: %s",
+			      UpnpGetErrorMessage(upnp_err)));
+    }
+#endif
+
+    pj_bzero(&upnp_mgr, sizeof(upnp_mgr));
+    upnp_err = UpnpInit2(param->if_name, (unsigned short)param->port);
+    if (upnp_err != UPNP_E_SUCCESS) {
```

Comment:
Q: has this been tested with pjsua restart scenario?

Reply:
Yes, this was the scenario that got my most attention because if pjsua is restarted when the network is down/changes, `libupnp` does create a lot of problems.

By default, `libupnp` creates its sockets in blocking mode, so a failed connection can take a long time before it times out. The solution for this, as described in the PR description above, is to disable blocking TCP connections when configuring libupnp. But then, even in non-blocking mode, `libupnp` sets a notoriously long timeout (30 secs) and the timeout is non-configurable so we need to modify the source (details in the desc above) to adjust this (I set it to 2 or 3 secs at most).

Still, the situation is not ideal because `UpnpFinish()` is a blocking function that will wait until every connection completes and all worker threads end, so it typically takes at least 4 secs to deinit:
```
10:21:48.904                 upnp.c  .UPnP deinitializing...
10:21:52.900                 upnp.c !Port mapping deletion action complete error: -207 (UPNP_E_TIMEDOUT)
10:21:52.900                 upnp.c !.UPnP deinitialized
```

But then, I imagine that UPnP should not be used a lot in dynamic/mobile network environments, so it's probably acceptable. I may be wrong here though.